### PR TITLE
Adding KV-Cache Support to Flux2.Klein Image Editing

### PR DESCRIPTION
**Pull request by amepas**
## Summary

Speeding up Image Editing on Flux2.Klein models by onboarding Flux2.Klein-9B-KV model variant. This is a separate model (same architecture, different weights) that was trained to support caching KV for reference images during image editing.

### Usage
Just adding the use_kv=True flag to the 9B model call will switch to the KV-cache model support. 

```python
python generate_flux2klein.py
  prompt="add a bunch of animals swimming in the water in front of the castle and get rid of all birds" \
  image_paths="['/path/to/img1.png', '/path/to/img2.png']" \
  use_kv=True
```
### Speed-Ups
Block-sizes can be tuned further for each setting. Using default settings still gives substantial improvements (using CP=2 on a TPU v7-2)

| Reference Images ($1024 \times 1024$) | Latency Speedup |
| :--- | :---: |
| 1 Image | **1.28x** |
| 2 Images | **1.32x** |
| 4 Images | **1.74x** |

### Visual Verification
Prompt: _change the painting so she is facing forward instead of looking over her shoulder_

Original:
image

Edited:
image

### Correctness 
KV-Cache implementation does not cause any regression or change in behavior on other flux2.klein pipelines.
Below numbers generated on v6-4.

- 4B T2I Smoke Test: 0.982 SSIM from baseline
- 9B T2I Smoke Test: 0.881 SSIM from baseline
- 4B Image Edit Smoke Test: 0.996 from baseline
- 9B Image Edit Smoke Test: 0.997 from baseline
- 9B KV-Cache Image Edit Smoke Test: 0.999 from baseline
- Diffusers KV-Cache Image Edit E2E Parity Test: 0.889 from PyTorch Diffusers

Note: KV-Cache requires a separate model checkpoint/training setup and is only supported on the 9B model.

### CODE REVIEW

File: src/maxdiffusion/pipelines/flux/flux2klein_pipeline.py [hunk 1]
```diff
@@ -27,18 +27,23 @@
 import numpy as np
 from flax.linen import partitioning as nn_partitioning
 
+from flax import nnx
```
[the remark this answers is not in the document]
Comment (amepas) [reply src/maxdiffusion/pipelines/flux/flux2klein_pipeline.py]: This is not correct. All packages were imported properly.

File: src/maxdiffusion/models/flux/vae/autoencoder_kl_flux2_nnx.py [hunk 2]
```diff
@@ -0,0 +1,799 @@
...
+    )
+
+  def __call__(self, x: jax.Array) -> jax.Array:
+    x = self.conv_in(x)
+    for block in self.down_blocks:
+      x = block(x)
+    x = self.mid_block(x)
+    x = self.conv_norm_out(x)
+    x = nnx.silu(x)
+    x = self.conv_out(x)
```
[the remark this answers is not in the document]
Comment (amepas) [reply src/maxdiffusion/models/flux/vae/autoencoder_kl_flux2_nnx.py:455]: This is an incorrect statement. nnx.silu is valid

File: src/maxdiffusion/models/flux/vae/autoencoder_kl_flux2_nnx.py [hunk 3]
```diff
@@ -0,0 +1,799 @@
...
+    else:
+      self.conv_shortcut = None
+
+  def __call__(self, x: jax.Array) -> jax.Array:
+    residual = self.conv_shortcut(x) if self.conv_shortcut is not None else x
+    h = self.norm1(x)
+    h = nnx.silu(h)
+    h = self.conv1(h)
+    h = self.norm2(h)
+    h = nnx.silu(h)
```
[the remark this answers is not in the document]
Comment (amepas) [reply src/maxdiffusion/models/flux/vae/autoencoder_kl_flux2_nnx.py:160]: This is an incorrect statement. nnx.silu is valid

File: src/maxdiffusion/models/flux/transformers/transformer_flux_flax.py [hunk 4]
```diff
@@ -1797,28 +2113,135 @@ def __call__(
...
-          temb=temb,
-          image_rotary_emb=concat_rotary_emb,
-          temb_mod_img=double_stream_mod_img,
-          temb_mod_txt=double_stream_mod_txt,
-      )
+    if kv_cache_mode == "extract" and num_ref_tokens > 0:
+      num_img_tokens = hidden_states.shape[1] - num_ref_tokens
+      ref_timestep = jnp.full_like(timestep_scaled, ref_fixed_timestep * 1000.0)
+      ref_temb = self.time_text_embed(ref_timestep, guidance_scaled, pooled_projections).astype(hidden_states.dtype)
+      ref_temb_silu = nnx.silu(ref_temb)
```
[the remark this answers is not in the document]
Comment (amepas) [reply src/maxdiffusion/models/flux/transformers/transformer_flux_flax.py]: This is an incorrect statement. nnx.silu is valid

File: src/maxdiffusion/generate_flux2klein.py [hunk 5]
```diff
@@ -79,8 +83,22 @@ def encode_prompt(prompt: str, snapshot_dir: str = None, repo_id: str = "black-f
 
   text_encoder_path = os.path.join(snapshot_dir, "text_encoder")
   tokenizer_path = os.path.join(snapshot_dir, "tokenizer")
-  if not os.path.exists(tokenizer_path):
-    tokenizer_path = text_encoder_path
+
+  if not os.path.exists(os.path.join(text_encoder_path, "config.json")) or not os.path.exists(tokenizer_path):
+    try:
+      fb_dir = snapshot_download(repo_id=repo_id, local_files_only=True)
```
[the remark this answers is not in the document]
Comment (amepas) [reply src/maxdiffusion/generate_flux2klein.py:93]: I think Gemini did not parse the file fully, snapshot_download is imported in this file.

File: src/maxdiffusion/models/flux/transformers/transformer_flux_flax.py [hunk 6]
```diff
@@ -1764,20 +2076,24 @@ def __call__(
...
-    if guidance is not None:
-      guidance = guidance * 1000.0
-    temb = self.time_text_embed(timestep, guidance, pooled_projections)
+    timestep_scaled = timestep * 1000.0
+    guidance_scaled = guidance * 1000.0 if guidance is not None else None
+    temb = self.time_text_embed(timestep_scaled, guidance_scaled, pooled_projections)
     temb = temb.astype(hidden_states.dtype)
 
-    temb_silu = jax.nn.silu(temb)
+    temb_silu = nnx.silu(temb)
```
[the remark this answers is not in the document]
Comment (amepas) [reply src/maxdiffusion/models/flux/transformers/transformer_flux_flax.py]: I'm not sure where this confusion came from for Gemini. nnx.silu is a valid attribute

File: src/maxdiffusion/models/flux/transformers/transformer_flux_flax.py [hunk 7]
```diff
@@ -1278,64 +1276,123 @@ def __call__(
...
+        kernel_init=nnx.with_partitioning(nnx.initializers.lecun_normal(), ("mlp", "embed")),
+        dtype=dtype,
+        param_dtype=weights_dtype,
+        rngs=rngs,
+    )
+
+  def __call__(self, x: jax.Array) -> jax.Array:
+    x = self.linear_in(x)
+    x1, x2 = jnp.split(x, 2, axis=-1)
+    hidden = nnx.silu(x1) * x2
```
[the remark this answers is not in the document]
Comment (amepas) [reply src/maxdiffusion/models/flux/transformers/transformer_flux_flax.py:1314]: Incorrect statement

File: src/maxdiffusion/models/flux/transformers/transformer_flux_flax.py [hunk 8]
```diff
@@ -1433,58 +1435,113 @@ def __call__(
...
+        k_txt = key_proj[:, :num_txt_tokens]
+        k_img = key_proj[:, num_txt_tokens:]
+        v_txt = value_proj[:, :num_txt_tokens]
+        v_img = value_proj[:, num_txt_tokens:]
+
+        k_all = jnp.concatenate([k_txt, k_ref, k_img], axis=1).reshape(B, -1, H * D)
+        v_all = jnp.concatenate([v_txt, v_ref, v_img], axis=1).reshape(B, -1, H * D)
+        q_all = query_proj.reshape(B, -1, H * D)
+
+        attn_output = self.attention_op.apply_attention(
```
Comment (eltsai) [inline src/maxdiffusion/models/flux/transformers/transformer_flux_flax.py:1522]: on line 1495 there is no `preserve_asymmetric_block_sizes=True`, I think if it is not set, the tuned `flash_block_sizes` is not passed in?

Comment (amepas) [reply src/maxdiffusion/models/flux/transformers/transformer_flux_flax.py:1522]: Yes, that is correct. Just amended the PR to account for that. SSIM scores did not change.

File: src/maxdiffusion/configs/base_flux2klein_9B.yml [hunk 9]
```diff
@@ -40,6 +41,9 @@ max_sequence_length: 512
 time_shift: True
 base_shift: 0.5
 max_shift: 1.15
+image_paths: []
+use_base2_exp: True
```
Comment (syhuang22) [inline src/maxdiffusion/configs/base_flux2klein_9B.yml:45]: This turns on base-2 softmax for the existing T2I path too, not just KV but the golden images weren't regenerated and the SSIM threshold went 0.75 -> 0.8 at the same time. 9B T2I at 0.881 vs >= 0.98 everywhere else looks like that drift.

File: src/maxdiffusion/models/flux/vae/autoencoder_kl_flux2_nnx.py [hunk 10]
```diff
@@ -0,0 +1,798 @@
...
+  def get_conv_kernel(key):
+    return jnp.array(pt_state_dict[key].transpose(2, 3, 1, 0), dtype=target_dtype)
+
+  def get_linear_kernel(key):
+    return jnp.array(pt_state_dict[key].T, dtype=target_dtype)
+
+  flat_state = dict(nnx.to_flat_state(nnx.state(nnx_vae, nnx.Param)))
+
+  def set_val(var, val):
+    var[...] = val
```
Comment (syhuang22) [inline src/maxdiffusion/models/flux/vae/autoencoder_kl_flux2_nnx.py]: The loader keeps norm weights in fp32, but the VAE they're written into was
built as bfloat16, so the assignment downcasts them straight back, making the fp32-norm path dead code.

Could we maybe try param_dtype=float32 on the norm layers (or drop is_norm if bf16 is intended), and use`val.astype(var.dtype)` here?

File: src/maxdiffusion/configs/base_flux2klein_9B.yml (hunk 9 above)
Comment (amepas) [reply src/maxdiffusion/configs/base_flux2klein_9B.yml:45]: If we didn't set use_base2_exp=True explicitly for all paths, there would be some errors for all paths (T2I and Image Editing) when we use other kernels so it was intended to be used for all paths. 

The SSIM threshold was changed to 0.8 just for consistency in all tests. (higher is better, so we actually made it more strict)

The reason for 9B T2I being 0.881 was a result of the optimizations in PR #458 that led to some re-arrangement of model computations. The T2I 9B SSIM did not change from PR #458, so adding the use_base2_exp flag did not affect the generated output.

Comment (amepas) [reply src/maxdiffusion/configs/base_flux2klein_9B.yml:45]: Marking this as resolved, but please let me know if you have any further questions!

File: src/maxdiffusion/models/flux/vae/autoencoder_kl_flux2_nnx.py (hunk 10 above)
Comment (amepas) [reply src/maxdiffusion/models/flux/vae/autoencoder_kl_flux2_nnx.py]: Good catch, we are now explicitly keeping the norm layers at float32 and using ```val.astype(var.dtype)``` in the quoted section now.  Re-ran all smoke tests and saw no drop in SSIM

**Review by syhuang22 — DISMISSED** [review]

**Review by eltsai — DISMISSED** [review]

File: src/maxdiffusion/tests/flux2klein_kv_pipeline_e2e_test.py [hunk 11]
```diff
@@ -0,0 +1,457 @@
...
+  return float(20 * np.log10(255.0 / np.sqrt(mse)))
+
+
+def compute_ssim(img1: Image.Image, img2: Image.Image) -> float:
+  arr1 = np.array(img1.convert("RGB"))
+  arr2 = np.array(img2.convert("RGB"))
+  return float(ssim(arr1, arr2, channel_axis=-1))
+
+
+def find_model_path():
```
Comment (mbohlool) [inline src/maxdiffusion/tests/flux2klein_kv_pipeline_e2e_test.py]: Nit: `find_model_path()` contains hardcoded machine paths (`/mnt/hyperdisk_weights/...`, `/mnt/data/...`). We can use `get_model_snapshot_dir()` (as in `edit_flux2klein_e2e_test.py`) for clean HF_HOME resolution.

File: src/maxdiffusion/tests/flux2klein_kv_pipeline_e2e_test.py [hunk 12]
```diff
@@ -0,0 +1,457 @@
...
+import torch
+
+try:
+  from diffusers import Flux2KleinKVPipeline
+except ImportError:
+  Flux2KleinKVPipeline = None
+
+import jax
+import jax.numpy as jnp
+import flax
```
Comment (mbohlool) [inline src/maxdiffusion/tests/flux2klein_kv_pipeline_e2e_test.py]: Nit: `import flax`, `import flax.linen as nn`, and `nn_partitioning` appear to be unused

**Review by mbohlool — COMMENTED** [review]
looks great with a couple of nit comments

**Review by mbohlool — DISMISSED** [review]

**Review by syhuang22 — APPROVED** [review]
